bubble_sort.rb

```
@@ -0,0 +1,51 @@
+# frozen_string_literal: true
+
+def bubble_sort(arr)
+  array_length = arr.length
+  return arr if array_length < 2
+
+  max_index = array_length - 2
+  swapped_elements = true
+
+  while swapped_elements
+    swapped_elements = false
+
+    (0..max_index).each do |i|
+      element = arr[i]
+      next_element = arr[i + 1]
+      next unless element > next_element
+
+      arr[i] = next_element
+      arr[i + 1] = element
+      swapped_elements = true
+    end
+  end
+
+  arr
+end
+
+def bubble_sort_by(arr)
+  if block_given?
+    swap = -1
+    while swap != 0
+      swap = 0
+      (0..(arr.length - 2)).each do |i|
+        if yield(arr[i], arr[i + 1]).positive?
+          arr[i], arr[i + 1] = arr[i + 1], arr[i]
+          swap += 1
+        end
+      end
+    end
+  else
+    p 'No block given'
+    # arr.sort
+  end
+
+  puts '--------------'
+  output = arr.join(', ')
+  puts "[#{output}]"
+end
```

bubble_sort_by(%w[hi hello hey]) { |left, right| left.length - right.length }) === %w[hi hey hello]
  
bubble_sort_by(%w[hi hello hey]) this should raise an error because no block of  code was given

---

1) bubble_sort_by sorts an array by accepting a block
     Failure/Error: expect(bubble_sort_by(unsorted) { |left, right| left.length - right.length }).to eq %w[hi hey hello]
     
       expected: ["hi", "hey", "hello"]
            got: nil
     
       (compared using ==)

  2) bubble_sort_by raises an error if block is not given
     Failure/Error: expect { bubble_sort_by(unsorted) }.to raise_error
       expected Exception but nothing was raised